### PR TITLE
feat(experimental): Add WebMCP Adapter

### LINKED ISSUE #1216: Adding WebMCP-compatible Support for Tools
Hi folks,

As discussed with @mattzcarey, sharing the fork branch that would add WebMCP compatibility for tools that are available via an MCP server: [https://github.com/alekrumkamp/agents/tree/feat/webmcp-adapter](https://github.com/alekrumkamp/agents/tree/feat/webmcp-adapter)

## How it works

1. registerWebMcp() connects to your MCP server
2. Discovers tools via tools/list
3. Registers each with navigator.modelContext.registerTool()
4. Chrome can now call your tools
5. Calls are relayed back to the MCP server via tools/call

## Usage
```javascript
import { registerWebMcp } from "agents/experimental/webmcp";

const handle = await registerWebMcp({ url: "/mcp" });
```

## Example where WebMCP tools get picked up by browser extension
Image

FYI @whoiskatrin

### PR DESCRIPTION
**Pull request by Muhammad-Bin-Ali**
## Add experimental WebMCP adapter

Bridges MCP server tools to Chrome's native `navigator.modelContext` API so browser-native AI agents can discover and call tools registered on a Cloudflare `McpAgent` — no extra infrastructure needed.

```ts
import { registerWebMcp } from "agents/experimental/webmcp";

const handle = await registerWebMcp({ url: "/mcp" });
// Chrome's AI can now call your MCP tools
```

### What's included

- **`agents/experimental/webmcp`** — New public export. Connects to an MCP server via the SDK's `StreamableHTTPClientTransport`, discovers tools, and registers them with `navigator.modelContext.registerTool()`.
- **`examples/webmcp/`** — Demo app with an `McpAgent` server and React client showing tool discovery, execution, and live refresh.
- **28 tests** running in Chromium via Playwright — covers tool execution, watch mode (SSE re-sync), error handling, edge cases, and known-bug regression tests.

### Design decisions

- Uses the official `@modelcontextprotocol/sdk` `Client` + `StreamableHTTPClientTransport` instead of a hand-rolled HTTP client. This gives us proper SSE parsing, reconnection with backoff, protocol version negotiation, session termination, and pagination for free.
- Supports `headers` (static) and `getHeaders` (per-request dynamic) for authenticated MCP endpoints.
- Gracefully no-ops when `navigator.modelContext` is unavailable (non-Chrome browsers).

#### Closes #1216

---

<a href="https://app.devin.ai/review/cloudflare/agents/pull/1222" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: packages/agents/src/experimental/webmcp.ts [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+  execute: (
+    input: Record<string, unknown>,
+    client: ModelContextClient
+  ) => Promise<unknown>;
+  annotations?: ModelContextToolAnnotations;
+}
+
+interface ModelContext {
+  registerTool(tool: ModelContextTool): void;
+  unregisterTool(name: string): void;
```
Comment (MiguelsPizza) [inline packages/agents/src/experimental/webmcp.ts:53]: Just a heads up, we use abort controllers to unregister tools now.

https://github.com/webmachinelearning/webmcp/issues/130
